### PR TITLE
[5.1] Better handling of invalid controller action reference

### PR DESCRIPTION
**Pull request by rkgrep**
see #9154

### CODE REVIEW

File: src/Illuminate/Routing/Route.php [hunk 1]
```diff
@@ -598,6 +599,8 @@ protected function replaceDefaults(array $parameters)
      *
      * @param  callable|array  $action
      * @return array
+     *
+     * @throws UnexpectedValueException
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Route.php]: Please use the FQCN.

Comment (GrahamCampbell) [inline src/Illuminate/Routing/Route.php]: Please use a slash at the start.

Comment (rkgrep) [inline src/Illuminate/Routing/Route.php]: Fixed

**Comment by ibrasho** [comment]
Why not:

```php
// Verify if provided "uses" property is valid Controller@action string
elseif (strpos($action['uses'], '@') !== false) {
    throw new UnexpectedValueException(sprintf('Invalid route action: %s', $action['uses']));
}
```

**Comment by rkgrep** [comment]
@ibrasho because double @ is not an expected value either.

**Comment by GrahamCampbell** [comment]
> // Verify if provided "uses" property is valid Controller@action string

They best way to do that is with a regular expression, though that would have an unnecessary performance kill.

**Comment by rkgrep** [comment]
I have just realized that a character after @ should be validated too. 

And I agree that regex is not an option.

**Comment by rkgrep** [comment]
Ok. I've made some tests. The first method is using a regex:
`!preg_match("/^(\\\\?)[a-zA-Z_\x7f-\xff](\\\\[a-zA-Z0-9_\x7f-\xff]|[a-zA-Z0-9_\x7f-\xff])*@[a-zA-Z_\x7f-\xff]([a-zA-Z0-9_\x7f-\xff]*)$/", $action['uses'])`

Which is quite slow, but it completely validates strings including all proper backslashes and valid PHP identifiers.

Second is just validate presense of a single @ at least at second position and a single non-@ character in the end
`substr_count($action['uses'], '@', 1) != 1 || substr($action['uses'], -1, 1) == '@'`
This does not give a proper validation though exceptions are thrown further on class construction. So we can possibly allow to pass something like `1@\` which will fail with a proper error description, but it won't trigger *Undefined offset* error - which is not so obvious.
Also it is only a bit slower than excluding all validation

regex_strings

Test source: https://gist.github.com/rkgrep/b87e7553a957ae0dde42

**Pull request merged**